File: src/middlewares/subscription.ts
import { Context, Markup } from 'telegraf';
import { config } from '../config';

export async function checkSubscription(ctx: Context, next: () => Promise<void>) {
    if (ctx.from?.id) {
        try {
            const member = await ctx.telegram.getChatMember(config.CHANNEL_ID, ctx.from.id);
            const statuses = ['member', 'administrator', 'creator'];

            if (statuses.includes(member.status)) {
                return next();
            } else {
                await ctx.reply(`Iltimos, botdan foydalanish uchun kanalimizga a'zo bo'ling:`,
                    Markup.inlineKeyboard([
                        [Markup.button.url('📢 Kanalga a\'zo bo\'lish', config.CHANNEL_URL)],
                        [Markup.button.callback('✅ Tekshirish', 'check_sub')]
                    ])
                );
            }
        } catch (error) {
            console.error('Subscription check error:', error);
            // If error occurs (e.g. bot not admin in channel), we might want to let them through or warn admin
            return next();
        }
    }
}

File: middlewares/subscription.py
from typing import Any, Awaitable, Callable, Dict
from aiogram import BaseMiddleware
from aiogram.types import Message, CallbackQuery, InlineKeyboardMarkup, InlineKeyboardButton
from config import settings

class SubscriptionMiddleware(BaseMiddleware):
    async def __call__(
        self,
        handler: Callable[[Message, Dict[str, Any]], Awaitable[Any]],
        event: Message,
        data: Dict[str, Any]
    ) -> Any:
        # Ignore non-messages and start command (start handler handles its own sub check)
        if not isinstance(event, Message) or (event.text and event.text.startswith('/start')):
            return await handler(event, data)

        user_id = event.from_user.id
        bot = data['bot']

        try:
            member = await bot.get_chat_member(chat_id=settings.channel_id, user_id=user_id)
            if member.status in ['member', 'administrator', 'creator']:
                return await handler(event, data)
            else:
                keyboard = InlineKeyboardMarkup(inline_keyboard=[
                    [InlineKeyboardButton(text="📢 Kanalga a'zo bo'lish", url=settings.channel_url)],
                    [InlineKeyboardButton(text="✅ Tekshirish", callback_data="check_sub")]
                ])
                await event.answer(
                    "Iltimos, botdan foydalanish uchun kanalimizga a'zo bo'ling:",
                    reply_markup=keyboard
                )
                return
        except Exception as e:
            print(f"Subscription check error: {e}")
            return await handler(event, data)
